compiler/rustc_traits/src/normalize_projection_ty.rs

```diff
@@ -40,6 +43,27 @@ fn normalize_projection_ty<'tcx>(
                 &mut obligations,
             );
             ocx.register_obligations(obligations);
+            // #112047: With projections and opaques, we are able to create opaques that
+            // are recursive (given some substitution of the opaque's type variables).
+            // In that case, we may only realize a cycle error when calling
+            // `normalize_erasing_regions` in mono.
+            if !ocx.infcx.next_trait_solver() {
+                let errors = ocx.select_where_possible();
+                if !errors.is_empty() {
+                    // Rustdoc may attempt to normalize type alias types which are not
+                    // well-formed. Rustdoc also normalizes types that are just not
+                    // well-formed, since we don't do as much HIR analysis (checking
+                    // that impl vars are constrained by the signature, for example).
```

Comment:
Does it do that for code that passes rustc? We are allowed to cause new errors in rustdoc if the code also wouldn't compile in rustc.

Reply:
>  a more graceful way than ICEing

Reply:
> We are allowed to cause new errors in rustdoc if the code also wouldn't compile in rustc

Well, the problem is that we need to do HIR analysis to make sure we don't have things like malformed and overlapping impl blocks too -- these are probably *not* what we want to "fix" in rustdoc, since we compile rustdoc with all features enabled and stuff?

> does rustdoc unwrap something instead of reporting an error? or where is the ICE coming from?

The ICE is coming from a failed projection in the `QueryNormalizer`. 

What happens is that we hit a cycle here:
https://github.com/rust-lang/rust/blob/0692db1a9082380e027f354912229dfd6af37e78/compiler/rustc_trait_selection/src/traits/project.rs#L1193-L1196

Which means we treat the projection as ambiguous and emit a projection goal with an infer var on the RHS. That causes us to hit this assertion with debug assertions enabled:
https://github.com/rust-lang/rust/blob/0692db1a9082380e027f354912229dfd6af37e78/compiler/rustc_traits/src/normalize_erasing_regions.rs#L50

Or just errors out with `NoSolution`, which causes `normalize_erasing_regions` to ICE.

Reply:
Ah... yea that makes sense sadly.

Tho Rustdoc is moving towards doing wf checks in the future, so this will get cleaned up